File: sdk/src/agui.rs
use crate::context::ContextAssembler;
use crate::error::LaserError;
use crate::laser::Laser;
use crate::provenance::AgentTopic;
use crate::types::ConversationId;
use laser_wire::agent::{
    AgentEnvelope, AgentId, AgentKind, OPERATION_REASONING, OPERATION_STATE_DELTA,
    OPERATION_STATE_SNAPSHOT, OPERATION_TASK, OPERATION_TOOL_ARGS, TaskState,
};
use laser_wire::content::ContentType;
use serde::Serialize;
use serde_json::Value as JsonValue;

/// An AG-UI protocol event, tagged by its SCREAMING_SNAKE `type` on the wire.
/// The events AGDX renders directly from the log: chat chunk streams become text
/// messages, reasoning chunk streams become reasoning messages, `tool_args`
/// chunk streams become tool calls (with the answering envelope as the tool
/// result), `status` task updates become run lifecycle, state events become
/// state events, and an error terminal becomes a run error.
#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(tag = "type")]
pub enum AgUiEvent {
    #[serde(rename = "RUN_STARTED")]
    RunStarted {
        #[serde(rename = "threadId")]
        thread_id: String,
        #[serde(rename = "runId")]
        run_id: String,
    },
    #[serde(rename = "RUN_FINISHED")]
    RunFinished {
        #[serde(rename = "threadId")]
        thread_id: String,
        #[serde(rename = "runId")]
        run_id: String,
    },
    #[serde(rename = "TEXT_MESSAGE_START")]
    TextMessageStart {
        #[serde(rename = "messageId")]
        message_id: String,
        role: String,
    },
    #[serde(rename = "TEXT_MESSAGE_CONTENT")]
    TextMessageContent {
        #[serde(rename = "messageId")]
        message_id: String,
        delta: String,
    },
    #[serde(rename = "TEXT_MESSAGE_END")]
    TextMessageEnd {
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "REASONING_MESSAGE_START")]
    ReasoningMessageStart {
        #[serde(rename = "messageId")]
        message_id: String,
        role: String,
    },
    #[serde(rename = "REASONING_MESSAGE_CONTENT")]
    ReasoningMessageContent {
        #[serde(rename = "messageId")]
        message_id: String,
        delta: String,
    },
    #[serde(rename = "REASONING_MESSAGE_END")]
    ReasoningMessageEnd {
        #[serde(rename = "messageId")]
        message_id: String,
    },
    #[serde(rename = "TOOL_CALL_START")]
    ToolCallStart {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        #[serde(rename = "toolCallName")]
        tool_call_name: String,
    },
    #[serde(rename = "TOOL_CALL_ARGS")]
    ToolCallArgs {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        delta: String,
    },
    #[serde(rename = "TOOL_CALL_END")]
    ToolCallEnd {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
    },
    #[serde(rename = "TOOL_CALL_RESULT")]
    ToolCallResult {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        content: String,
    },
    #[serde(rename = "STATE_SNAPSHOT")]
    StateSnapshot { snapshot: JsonValue },
    #[serde(rename = "STATE_DELTA")]
    StateDelta { delta: JsonValue },
    #[serde(rename = "RUN_ERROR")]
    RunError { message: String },
}

impl Laser {
    /// Publish an AG-UI state snapshot: the full shared state as a
    /// `state_snapshot` event (`agdx.ct = json`). Replaying a snapshot plus the
    /// later deltas reconstructs the state at any historical offset, the
    /// log-native form of AG-UI's STATE_SNAPSHOT/STATE_DELTA, over Iggy's own
    /// transport rather than SSE.
    pub async fn publish_state_snapshot(
        &self,
        topic: AgentTopic<'static>,
        source: AgentId,
        conversation: ConversationId,
        state: &JsonValue,
    ) -> Result<(), LaserError> {
        let body =
            serde_json::to_vec(state).map_err(|error| LaserError::Codec(error.to_string()))?;
        self.agdx(topic, source, conversation.into())
            .emit(body)
            .with_operation(OPERATION_STATE_SNAPSHOT)
            .content_type(ContentType::Json)
            .send()
            .await?;
        Ok(())
    }

    /// Publish an AG-UI state delta: an RFC 6902 JSON Patch document as a
    /// `state_delta` event (`agdx.ct = json`).
    pub async fn publish_state_delta(
        &self,
        topic: AgentTopic<'static>,
        source: AgentId,
        conversation: ConversationId,
        patch: &JsonValue,
    ) -> Result<(), LaserError> {
        let body =
            serde_json::to_vec(patch).map_err(|error| LaserError::Codec(error.to_string()))?;
        self.agdx(topic, source, conversation.into())
            .emit(body)
            .with_operation(OPERATION_STATE_DELTA)
            .content_type(ContentType::Json)
            .send()
            .await?;
        Ok(())
    }

    /// Reconstruct shared state by replaying `conversation`'s `state_snapshot` /
    /// `state_delta` events on `topic`: take the latest snapshot, then apply
    /// every delta after it (RFC 6902). `None` until a snapshot exists.
    pub async fn reconstruct_state(
        &self,
        conversation: ConversationId,
        topic: AgentTopic<'static>,
    ) -> Result<Option<JsonValue>, LaserError> {
        let messages = ContextAssembler::builder()
            .conversation_id(conversation)
            .topics(vec![topic])
            .build()
            .assemble(self)
            .await?;
        let mut state: Option<JsonValue> = None;
        for message in &messages {
            let Some(envelope) = &message.envelope else {
                continue;
            };
            if envelope.kind != AgentKind::Event {
                continue;
            }
            match envelope.operation.as_deref() {
                Some(OPERATION_STATE_SNAPSHOT) => {
                    state = Some(
                        serde_json::from_slice(&envelope.body)
                            .map_err(|error| LaserError::Codec(error.to_string()))?,
                    );
                }
                Some(OPERATION_STATE_DELTA) => {
                    if let Some(document) = state.as_mut() {
                        let patch: json_patch::Patch = serde_json::from_slice(&envelope.body)
                            .map_err(|error| LaserError::Codec(error.to_string()))?;
                        json_patch::patch(document, &patch)
                            .map_err(|error| LaserError::Invalid(error.to_string()))?;
                    }
                }
                _ => {}
            }
        }
        Ok(state)
    }

    /// Render `conversation` on `topic` as AG-UI events by reading the log:
    /// chat/reasoning chunk streams become `TEXT_MESSAGE_*` events, state events
    /// become `STATE_SNAPSHOT`/`STATE_DELTA`, an error terminal becomes
    /// `RUN_ERROR`. Log-native, replayable, over Iggy rather than SSE.
    pub async fn agui_events(
        &self,
        conversation: ConversationId,
        topic: AgentTopic<'static>,
    ) -> Result<Vec<AgUiEvent>, LaserError> {
        let messages = ContextAssembler::builder()
            .conversation_id(conversation)
            .topics(vec![topic])
            .build()
            .assemble(self)
            .await?;
        // The chunk-stream purpose rides only the opening chunk (sequence 0), so
        // track each channel's kind as the stream opens and reuse it for the
        // later chunks (a terminal chunk carries no purpose).
        let mut channels: std::collections::HashMap<String, ChunkKind> =
            std::collections::HashMap::new();
        let mut events = Vec::new();
        for message in &messages {
            let Some(envelope) = &message.envelope else {
                continue;
            };
            if envelope.kind == AgentKind::Chunk {
                let id = envelope
                    .channel
                    .map(|channel| channel.to_string())
                    .unwrap_or_default();
                let kind = if envelope.sequence == Some(0) {
                    let kind = chunk_kind_of(envelope);
                    channels.insert(id.clone(), kind);
                    kind
                } else {
                    channels.get(&id).copied().unwrap_or(ChunkKind::Chat)
                };
                events.extend(chunk_to_agui(envelope, kind));
            } else {
                events.extend(envelope_to_agui(envelope));
            }
        }
        Ok(events)
    }
}

// The chunk-stream purpose the opening chunk declares decides which AG-UI
// message family the stream renders as.
#[derive(Clone, Copy)]
enum ChunkKind {
    Chat,
    Reasoning,
    ToolArgs,
}

// The kind an opening chunk's purpose declares (mid-stream chunks carry none).
fn chunk_kind_of(envelope: &AgentEnvelope) -> ChunkKind {
    match envelope.operation.as_deref() {
        Some(OPERATION_REASONING) => ChunkKind::Reasoning,
        Some(OPERATION_TOOL_ARGS) => ChunkKind::ToolArgs,
        _ => ChunkKind::Chat,
    }
}

/// Translate one non-chunk AGDX envelope into the AG-UI events it represents
/// (chunks are handled by [`Laser::agui_events`], which threads the per-channel
/// stream kind).
fn envelope_to_agui(envelope: &AgentEnvelope) -> Vec<AgUiEvent> {
    match envelope.kind {
        AgentKind::Status if envelope.operation.as_deref() == Some(OPERATION_TASK) => {
            // A task lifecycle update: submitted opens the run, a terminal state
            // closes it. threadId = conversation, runId = correlation.
            let thread_id = envelope.conversation.to_string();
            let run_id = envelope
                .correlation
                .map(|correlation| correlation.to_string())
                .unwrap_or_default();
            match envelope.task_state {
                Some(TaskState::Submitted) => vec![AgUiEvent::RunStarted { thread_id, run_id }],
                Some(state) if state.is_terminal() => {
                    vec![AgUiEvent::RunFinished { thread_id, run_id }]
                }
                _ => Vec::new(),
            }
        }
        // A response/error carrying `tool` is a tool result.
        AgentKind::Response | AgentKind::Error if envelope.tool.is_some() => {
            let tool_call_id = envelope
                .correlation
                .map(|correlation| correlation.to_string())
                .unwrap_or_default();
            vec![AgUiEvent::ToolCallResult {
                tool_call_id,
                content: String::from_utf8_lossy(&envelope.body).into_owned(),
            }]
        }
        AgentKind::Error => vec![AgUiEvent::RunError {
            message: String::from_utf8_lossy(&envelope.body).into_owned(),
        }],
        AgentKind::Event => match envelope.operation.as_deref() {
            Some(OPERATION_STATE_SNAPSHOT) => serde_json::from_slice(&envelope.body)
                .map(|snapshot| vec![AgUiEvent::StateSnapshot { snapshot }])
                .unwrap_or_default(),
            Some(OPERATION_STATE_DELTA) => serde_json::from_slice(&envelope.body)
                .map(|delta| vec![AgUiEvent::StateDelta { delta }])
                .unwrap_or_default(),
            _ => Vec::new(),
        },
        _ => Vec::new(),
    }
}

// Render a chunk of a chat / reasoning / tool_args stream as its AG-UI message
// family. The stream's purpose is declared on the opening chunk (sequence 0),
// so a reader tracks it per channel. Here each chunk is self-describing enough
// because the opening chunk carries the purpose and later chunks of a known
// channel reuse it. The opening chunk decides the kind, and the consumer threads it.
fn chunk_to_agui(envelope: &AgentEnvelope, kind: ChunkKind) -> Vec<AgUiEvent> {
    let id = envelope
        .channel
        .map(|channel| channel.to_string())
        .unwrap_or_default();
    let body = String::from_utf8_lossy(&envelope.body).into_owned();
    let opening = envelope.sequence == Some(0);
    let mut events = Vec::new();
    match kind {
        ChunkKind::Chat => {
            if opening {
                events.push(AgUiEvent::TextMessageStart {
                    message_id: id.clone(),
                    role: "assistant".to_owned(),
                });
            }
            if !body.is_empty() {
                events.push(AgUiEvent::TextMessageContent {
                    message_id: id.clone(),
                    delta: body,
                });
            }
            if envelope.last {
                events.push(AgUiEvent::TextMessageEnd { message_id: id });
            }
        }
        ChunkKind::Reasoning => {
            if opening {
                events.push(AgUiEvent::ReasoningMessageStart {
                    message_id: id.clone(),
                    role: "reasoning".to_owned(),
                });
            }
            if !body.is_empty() {
                events.push(AgUiEvent::ReasoningMessageContent {
                    message_id: id.clone(),
                    delta: body,
                });
            }
            if envelope.last {
                events.push(AgUiEvent::ReasoningMessageEnd { message_id: id });
            }
        }
        ChunkKind::ToolArgs => {
            if opening {
                events.push(AgUiEvent::ToolCallStart {
                    tool_call_id: id.clone(),
                    tool_call_name: envelope.tool.clone().unwrap_or_default(),
                });
            }
            if !body.is_empty() {
                events.push(AgUiEvent::ToolCallArgs {
                    tool_call_id: id.clone(),
                    delta: body,
                });
            }
            if envelope.last {
                events.push(AgUiEvent::ToolCallEnd { tool_call_id: id });
            }
        }
    }
    events
}

File: foreign/typescript/src/bridges/agui.ts
import { CodecError, InvalidError } from "../client/errors.js"
import type { Laser } from "../client/laser.js"
import type { AgentId, ConversationId } from "../types/ids.js"
import {
  AgentKind,
  OPERATION_REASONING,
  OPERATION_STATE_DELTA,
  OPERATION_STATE_SNAPSHOT,
  OPERATION_TASK,
  OPERATION_TOOL_ARGS,
  taskStateIsTerminal,
  type AgentEnvelope
} from "../wire/agent.js"
import { ContentType } from "../wire/content.js"

export type AgUiEvent =
  | { readonly type: "RUN_STARTED"; readonly threadId: string; readonly runId: string }
  | { readonly type: "RUN_FINISHED"; readonly threadId: string; readonly runId: string }
  | { readonly type: "TEXT_MESSAGE_START"; readonly messageId: string; readonly role: string }
  | { readonly type: "TEXT_MESSAGE_CONTENT"; readonly messageId: string; readonly delta: string }
  | { readonly type: "TEXT_MESSAGE_END"; readonly messageId: string }
  | { readonly type: "REASONING_MESSAGE_START"; readonly messageId: string; readonly role: string }
  | {
      readonly type: "REASONING_MESSAGE_CONTENT"
      readonly messageId: string
      readonly delta: string
    }
  | { readonly type: "REASONING_MESSAGE_END"; readonly messageId: string }
  | { readonly type: "TOOL_CALL_START"; readonly toolCallId: string; readonly toolCallName: string }
  | { readonly type: "TOOL_CALL_ARGS"; readonly toolCallId: string; readonly delta: string }
  | { readonly type: "TOOL_CALL_END"; readonly toolCallId: string }
  | { readonly type: "TOOL_CALL_RESULT"; readonly toolCallId: string; readonly content: string }
  | { readonly type: "STATE_SNAPSHOT"; readonly snapshot: unknown }
  | { readonly type: "STATE_DELTA"; readonly delta: unknown }
  | { readonly type: "RUN_ERROR"; readonly message: string }

type ChunkKind = "chat" | "reasoning" | "toolArgs"

function jsonBytes(value: unknown, operation: string): Uint8Array {
  try {
    const encoded: unknown = JSON.stringify(value)
    if (typeof encoded !== "string") throw new Error("value is not JSON serializable")
    return new TextEncoder().encode(encoded)
  } catch (cause) {
    throw new CodecError(`cannot encode ${operation}`, "agui", operation, { cause })
  }
}

function parseJson(bytes: Uint8Array, operation: string): unknown {
  try {
    return JSON.parse(new TextDecoder("utf-8", { fatal: true }).decode(bytes)) as unknown
  } catch (cause) {
    throw new CodecError(`cannot decode ${operation}`, "agui", operation, { cause })
  }
}

function decodePointer(path: string): readonly string[] {
  if (path === "") return []
  if (!path.startsWith("/")) throw new InvalidError(`invalid JSON Pointer \`${path}\``)
  return path
    .slice(1)
    .split("/")
    .map((part) => {
      if (/~(?:[^01]|$)/.test(part)) throw new InvalidError(`invalid JSON Pointer \`${path}\``)
      return part.replaceAll("~1", "/").replaceAll("~0", "~")
    })
}

function arrayIndex(token: string, length: number, allowEnd: boolean): number {
  if (token === "-" && allowEnd) return length
  if (!/^(?:0|[1-9][0-9]*)$/.test(token)) {
    throw new InvalidError(`invalid JSON Patch array index \`${token}\``)
  }
  const index = Number(token)
  const maximum = allowEnd ? length : length - 1
  if (!Number.isSafeInteger(index) || index < 0 || index > maximum) {
    throw new InvalidError(`JSON Patch array index \`${token}\` is out of bounds`)
  }
  return index
}

function objectOf(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new InvalidError(`JSON Patch path \`${path}\` does not address a container`)
  }
  return value as Record<string, unknown>
}

function parentAt(document: unknown, tokens: readonly string[], path: string): [unknown, string] {
  if (tokens.length === 0) throw new InvalidError("the document root has no parent")
  let current = document
  for (const token of tokens.slice(0, -1)) {
    if (Array.isArray(current)) {
      current = current[arrayIndex(token, current.length, false)]
    } else {
      const object = objectOf(current, path)
      if (!(token in object)) throw new InvalidError(`JSON Patch path \`${path}\` does not exist`)
      current = object[token]
    }
  }
  return [current, tokens.at(-1) ?? ""]
}

function valueAt(document: unknown, path: string): unknown {
  const tokens = decodePointer(path)
  let current = document
  for (const token of tokens) {
    if (Array.isArray(current)) current = current[arrayIndex(token, current.length, false)]
    else {
      const object = objectOf(current, path)
      if (!(token in object)) throw new InvalidError(`JSON Patch path \`${path}\` does not exist`)
      current = object[token]
    }
  }
  return current
}

function addValue(document: unknown, path: string, value: unknown): unknown {
  const tokens = decodePointer(path)
  if (tokens.length === 0) return value
  const [parent, token] = parentAt(document, tokens, path)
  if (Array.isArray(parent)) parent.splice(arrayIndex(token, parent.length, true), 0, value)
  else objectOf(parent, path)[token] = value
  return document
}

function removeValue(
  document: unknown,
  path: string
): { readonly document: unknown; readonly value: unknown } {
  const tokens = decodePointer(path)
  if (tokens.length === 0) return { document: null, value: document }
  const [parent, token] = parentAt(document, tokens, path)
  if (Array.isArray(parent)) {
    const index = arrayIndex(token, parent.length, false)
    const value: unknown = parent[index]
    parent.splice(index, 1)
    return { document, value }
  }
  const object = objectOf(parent, path)
  if (!(token in object)) throw new InvalidError(`JSON Patch path \`${path}\` does not exist`)
  const value = object[token]
  Reflect.deleteProperty(object, token)
  return { document, value }
}

function equalJson(left: unknown, right: unknown): boolean {
  if (Object.is(left, right)) return true
  if (Array.isArray(left) && Array.isArray(right)) {
    return (
      left.length === right.length && left.every((value, index) => equalJson(value, right[index]))
    )
  }
  if (
    typeof left === "object" &&
    left !== null &&
    !Array.isArray(left) &&
    typeof right === "object" &&
    right !== null &&
    !Array.isArray(right)
  ) {
    const a = left as Readonly<Record<string, unknown>>
    const b = right as Readonly<Record<string, unknown>>
    const keys = Object.keys(a)
    return (
      keys.length === Object.keys(b).length &&
      keys.every((key) => key in b && equalJson(a[key], b[key]))
    )
  }
  return false
}

function patchObject(value: unknown): Readonly<Record<string, unknown>> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new InvalidError("each JSON Patch operation must be an object")
  }
  return value as Readonly<Record<string, unknown>>
}

export function applyJsonPatch(document: unknown, patch: unknown): unknown {
  if (!Array.isArray(patch)) throw new InvalidError("a JSON Patch document must be an array")
  let result = structuredClone(document)
  for (const raw of patch) {
    const operation = patchObject(raw)
    const op = operation["op"]
    const path = operation["path"]
    if (typeof op !== "string" || typeof path !== "string") {
      throw new InvalidError("a JSON Patch operation requires string `op` and `path`")
    }
    switch (op) {
      case "add":
        if (!("value" in operation)) throw new InvalidError("JSON Patch add requires `value`")
        result = addValue(result, path, structuredClone(operation["value"]))
        break
      case "remove":
        result = removeValue(result, path).document
        break
      case "replace":
        if (!("value" in operation)) throw new InvalidError("JSON Patch replace requires `value`")
        valueAt(result, path)
        result = removeValue(result, path).document
        result = addValue(result, path, structuredClone(operation["value"]))
        break
      case "move": {
        const from = operation["from"]
        if (typeof from !== "string") throw new InvalidError("JSON Patch move requires `from`")
        if (path.startsWith(`${from}/`))
          throw new InvalidError("JSON Patch cannot move a value into its child")
        const removed = removeValue(result, from)
        result = addValue(removed.document, path, removed.value)
        break
      }
      case "copy": {
        const from = operation["from"]
        if (typeof from !== "string") throw new InvalidError("JSON Patch copy requires `from`")
        result = addValue(result, path, structuredClone(valueAt(result, from)))
        break
      }
      case "test":
        if (!("value" in operation)) throw new InvalidError("JSON Patch test requires `value`")
        if (!equalJson(valueAt(result, path), operation["value"])) {
          throw new InvalidError(`JSON Patch test failed at \`${path}\``)
        }
        break
      default:
        throw new InvalidError(`unknown JSON Patch operation \`${op}\``)
    }
  }
  return result
}

function chunkKind(envelope: AgentEnvelope): ChunkKind {
  if (envelope.operation === OPERATION_REASONING) return "reasoning"
  if (envelope.operation === OPERATION_TOOL_ARGS) return "toolArgs"
  return "chat"
}

function chunkEvents(envelope: AgentEnvelope, kind: ChunkKind): readonly AgUiEvent[] {
  const id = envelope.channel?.toString() ?? ""
  const body = new TextDecoder().decode(envelope.body)
  const opening = envelope.sequence === 0n
  const events: AgUiEvent[] = []
  if (kind === "chat") {
    if (opening) events.push({ type: "TEXT_MESSAGE_START", messageId: id, role: "assistant" })
    if (body.length > 0) events.push({ type: "TEXT_MESSAGE_CONTENT", messageId: id, delta: body })
    if (envelope.last) events.push({ type: "TEXT_MESSAGE_END", messageId: id })
  } else if (kind === "reasoning") {
    if (opening) events.push({ type: "REASONING_MESSAGE_START", messageId: id, role: "reasoning" })
    if (body.length > 0)
      events.push({ type: "REASONING_MESSAGE_CONTENT", messageId: id, delta: body })
    if (envelope.last) events.push({ type: "REASONING_MESSAGE_END", messageId: id })
  } else {
    if (opening)
      events.push({ type: "TOOL_CALL_START", toolCallId: id, toolCallName: envelope.tool ?? "" })
    if (body.length > 0) events.push({ type: "TOOL_CALL_ARGS", toolCallId: id, delta: body })
    if (envelope.last) events.push({ type: "TOOL_CALL_END", toolCallId: id })
  }
  return events
}

export function envelopeToAgUi(envelope: AgentEnvelope): readonly AgUiEvent[] {
  if (envelope.kind === AgentKind.Status && envelope.operation === OPERATION_TASK) {
    const threadId = envelope.conversation.toString()
    const runId = envelope.correlation?.toString() ?? ""
    if (envelope.taskState?.kind === "known" && envelope.taskState.name === "Submitted") {
      return [{ type: "RUN_STARTED", threadId, runId }]
    }
    if (envelope.taskState !== undefined && taskStateIsTerminal(envelope.taskState)) {
      return [{ type: "RUN_FINISHED", threadId, runId }]
    }
    return []
  }
  if (
    (envelope.kind === AgentKind.Response || envelope.kind === AgentKind.Error) &&
    envelope.tool !== undefined
  ) {
    return [
      {
        type: "TOOL_CALL_RESULT",
        toolCallId: envelope.correlation?.toString() ?? "",
        content: new TextDecoder().decode(envelope.body)
      }
    ]
  }
  if (envelope.kind === AgentKind.Error) {
    return [{ type: "RUN_ERROR", message: new TextDecoder().decode(envelope.body) }]
  }
  if (envelope.kind === AgentKind.Event && envelope.operation === OPERATION_STATE_SNAPSHOT) {
    try {
      return [{ type: "STATE_SNAPSHOT", snapshot: parseJson(envelope.body, "state snapshot") }]
    } catch {
      return []
    }
  }
  if (envelope.kind === AgentKind.Event && envelope.operation === OPERATION_STATE_DELTA) {
    try {
      return [{ type: "STATE_DELTA", delta: parseJson(envelope.body, "state delta") }]
    } catch {
      return []
    }
  }
  return []
}

export function envelopesToAgUi(envelopes: readonly AgentEnvelope[]): readonly AgUiEvent[] {
  const channels = new Map<string, ChunkKind>()
  const events: AgUiEvent[] = []
  for (const envelope of envelopes) {
    if (envelope.kind !== AgentKind.Chunk) {
      events.push(...envelopeToAgUi(envelope))
      continue
    }
    const id = envelope.channel?.toString() ?? ""
    const kind = envelope.sequence === 0n ? chunkKind(envelope) : (channels.get(id) ?? "chat")
    if (envelope.sequence === 0n) channels.set(id, kind)
    events.push(...chunkEvents(envelope, kind))
  }
  return events
}

export async function publishStateSnapshot(
  laser: Laser,
  topic: string,
  source: AgentId,
  conversation: ConversationId,
  state: unknown
): Promise<void> {
  await laser
    .agdx(topic, source, conversation)
    .emit(jsonBytes(state, "state snapshot"))
    .withOperation(OPERATION_STATE_SNAPSHOT)
    .contentType(ContentType.Json)
    .send()
}

export async function publishStateDelta(
  laser: Laser,
  topic: string,
  source: AgentId,
  conversation: ConversationId,
  patch: unknown
): Promise<void> {
  if (!Array.isArray(patch))
    throw new InvalidError("an AG-UI state delta must be a JSON Patch array")
  await laser
    .agdx(topic, source, conversation)
    .emit(jsonBytes(patch, "state delta"))
    .withOperation(OPERATION_STATE_DELTA)
    .contentType(ContentType.Json)
    .send()
}

export async function reconstructState(
  laser: Laser,
  conversation: ConversationId,
  topic: string
): Promise<unknown> {
  const messages = await laser.context(conversation).fetch([topic], Number.MAX_SAFE_INTEGER)
  let state: unknown = undefined
  for (const message of messages) {
    const envelope = message.envelope
    if (envelope?.kind !== AgentKind.Event) continue
    if (envelope.operation === OPERATION_STATE_SNAPSHOT) {
      state = parseJson(envelope.body, "state snapshot")
    } else if (envelope.operation === OPERATION_STATE_DELTA && state !== undefined) {
      state = applyJsonPatch(state, parseJson(envelope.body, "state delta"))
    }
  }
  return state
}

export async function aguiEvents(
  laser: Laser,
  conversation: ConversationId,
  topic: string
): Promise<readonly AgUiEvent[]> {
  const messages = await laser.context(conversation).fetch([topic], Number.MAX_SAFE_INTEGER)
  return envelopesToAgUi(
    messages.flatMap((message) => (message.envelope === undefined ? [] : [message.envelope]))
  )
}
